### PR TITLE
Fix nxService crashing when service does not exist

### LINKED ISSUE #30: nxService - Crashes when configuration refers to a service that does not exist
#### Details of the scenario you tried and the problem that is occurring
I just tested nxService again after https://github.com/Azure/nxtools/issues/25 got patched and closed.   Unfortunately my testing shows this still does not work properly in all situations, and crashes during consistence check.
However, I suspect that I've actually discovered a new bug.  It isn't exactly the same as issue #25

It appears that the nxService implementation is not correctly handling the case where the nxService configuration is referring to a service that does not exist on the target system;  the 'rhnsd' service does not exist on the Ubuntu system on which I ran the compliance check.  Usually that service would only exist on RHEL systems. 
If I try a different configuration that refers to a service that does exist on the Ubuntu system, such as 'auditd', then the error stack does not get output during the consistency check.

#### Verbose logs showing the problem
Error: Failed to Run Consistency for 'RHEL7-CISv311-0_0_6' Error : Index operation failed; the array index evaluated to null.
InvokeDscResource() failed. errorMessage:The PowerShell DSC resource with className:nxService moduleName:nxtools moduleVersion:1.2.0 threw an exception: System.Management.Automation.RuntimeException: Index operation failed; the array index evaluated to null.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0)
   at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(ScriptBlockClauseToInvoke clauseToInvoke, Boolean createLocalScope, Dictionary`2 functionsToDefine, List`1 variablesToDefine, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(Boolean createLocalScope, Dictionary`2 functionsToDefine, List`1 variablesToDefine, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeWithPipe(Boolean useLocalScope, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Boolean propagateAllExceptionsToTop, List`1 variablesToDefine, Dictionary`2 functionsToDefine, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeAsMemberFunctionT[T](Object instance, Object[] args)
   at System.Management.Automation.Internal.ScriptBlockMemberMethodWrapper.InvokeHelperT[T](Object instance, Object sessionStateInternal, Object[] args)
   at CallSite.Target(Closure , CallSite , Object , Object )
   at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
   at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame), while running the Get-DscResource functionality
The PowerShell DSC resource with className:nxService moduleName:nxtools moduleVersion:1.2.0 threw an exception: System.Management.Automation.RuntimeException: Index operation failed; the array index evaluated to null.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0)
   at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(ScriptBlockClauseToInvoke clauseToInvoke, Boolean createLocalScope, Dictionary`2 functionsToDefine, List`1 variablesToDefine, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(Boolean createLocalScope, Dictionary`2 functionsToDefine, List`1 variablesToDefine, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeWithPipe(Boolean useLocalScope, ErrorHandlingBehavior errorHandlingBehavior, Object dollarUnder, Object input, Object scriptThis, Pipe outputPipe, InvocationInfo invocationInfo, Boolean propagateAllExceptionsToTop, List`1 variablesToDefine, Dictionary`2 functionsToDefine, Object[] args)
   at System.Management.Automation.ScriptBlock.InvokeAsMemberFunctionT[T](Object instance, Object[] args)
   at System.Management.Automation.Internal.ScriptBlockMemberMethodWrapper.InvokeHelperT[T](Object instance, Object sessionStateInternal, Object[] args)
   at CallSite.Target(Closure , CallSite , Object , Object )
   at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
   at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame), while running the Get-DscResource functionality

Name                Path
----                ----
RHEL7-CISv311-0_0_6 /home/azops.eric.ehret/azam/RHEL7/Packages/RHEL7-CISv311-0_0_6.zip
Write-Error: Job e1f68492-6185-4c28-9d59-44e62bab7aa8 : MIResult: 1 Error Message: Index operation failed; the array index evaluated to null. Message ID: OMI:MI_Result:1 Error Category: 0 Error Code: 1 Error Type:
Write-Error: Job e1f68492-6185-4c28-9d59-44e62bab7aa8 : MIResult: 1 Error Message: InvokeDscResource() failed. errorMessage:The PowerShell DSC resource with className:nxService moduleName:nxtools moduleVersion:1.2.0 threw
an exception: System.Management.Automation.RuntimeException: Index operation failed; the array index evaluated to null.    at
System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)    at
System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)    at
System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)    at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame)    at
System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0)    at System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(ScriptBlockClauseTo
Write-Error: Job e1f68492-6185-4c28-9d59-44e62bab7aa8 : MIResult: 1 Error Message: The PowerShell DSC resource with className:nxService moduleName:nxtools moduleVersion:1.2.0 threw an exception:
System.Management.Automation.RuntimeException: Index operation failed; the array index evaluated to null.    at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext
funcContext, Exception exception)    at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)    at
System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
at System.Management.Automation.Interpreter.Interpreter.Run(InterpretedFrame frame)    at System.Management.Automation.Interpreter.LightLambda.RunVoid1[T0](T0 arg0)    at
System.Management.Automation.ScriptBlock.InvokeWithPipeImpl(ScriptBlockClauseToInvoke clauseToInvoke, Boolean createLoca
@{additionalProperties=System.Object[]; assignmentName=RHEL7-CISv311-0_0_6; complianceStatus=False; endTime=09/09/2023 23:42:17; jobId=e1f68492-6185-4c28-9d59-44e62bab7aa8; operationtype=Consistency; resources=System.Object[]; startTime=09/09/2023 23:42:17}

#### Suggested solution to the issue
I suspect that the correct thing to do in all cases where the named service does not exist would be to return reasons indicating that the service does not exist.
Less clear to me is whether this condition should always be considered 'non compliant'.  
Maybe it could be argued that if the service is supposed to be disabled then non-existence might be considered compliant as well.   I leave that to others to debate.   But I strongly believe it should not just crash without returning some kind of compliance result.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
    nxService rhnsd_disabled
    {
        Name = "rhnsd"
        Enabled = $false
        State = "Stopped"
    }
```

#### The operating system the target node is running
Crash occurred on Ubuntu 18 system that is being used to author the configurations.

#### Version and build of PowerShell the target node is running
7.3.6

#### Version of the DSC module that was used
1.2.0

### PR DESCRIPTION
**Pull request by gurrajatwal**
#### Pull Request (PR) description

The nxService resource crashes when auditing a service that does not exist. The if statement that was supposed to handle this case had the wrong condition.

#### This Pull Request (PR) fixes the following issues

- Fixes #30 

#### Task list

- [x] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [x] Resource documentation added/updated in README.md.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [ ] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

### CODE REVIEW

File: source/Classes/1.DscResources/07.nxService.ps1 [hunk 1]
```diff
@@ -40,7 +40,7 @@ class nxService
         $currentState.State = $nxService.State
         $currentState.Controller = $this.Controller
 
-        if (-not $currentState)
```
Comment (kwirkykat) [inline source/Classes/1.DscResources/07.nxService.ps1:43]: This check should happen before the info from the service is retrieved in the lines above. (It should be a check on the retrieved $nxService rather than $currentState too)
Also $currentState.Name is not a boolean so calling it as a boolean can have some unpredictable behavior.
This should be something like $null -eq $currentState.Name or some boolean expression like that

File: source/Classes/1.DscResources/07.nxService.ps1 [hunk 2]
```diff
@@ -40,7 +40,7 @@ class nxService
...
+        if (-not $currentState.Name)
         {
             # Silently return if the service does not exist
             Write-Warning -Message ('Service ''{0}'' could not be found.' -f $this.Name)
```
Comment (kwirkykat) [inline source/Classes/1.DscResources/07.nxService.ps1:46]: I know you didn't edit this part, but can we not return just a completely blank nxService in this case? (line 47)
Can we at least return a Reason that says the service does not exist or something?

File: tests/Unit/Classes/DscResources/nxService.tests.ps1 [hunk 3]
```diff
@@ -5,9 +5,7 @@ $script:testService = "myTestService.service" # Mock service for testing
 Describe "nxService resource for managing services on a Linux node" {
     BeforeAll {
         Mock -ModuleName 'nxtools' -CommandName 'Invoke-NativeCommand' -ParameterFilter {
-            $expected = @('is-enabled', $testService)
-            $diff = Compare-Object $Parameters $expected
-            return $Executable -eq "systemctl" -and $diff.Count -eq 0
+            return $Executable -eq "systemctl" -and $Parameters -contains "is-enabled"
```
Comment (kwirkykat) [inline tests/Unit/Classes/DscResources/nxService.tests.ps1:8]: Taking out the exact compare makes this check different? Before $Parameters had to be exactly @($testService, is-enabled)
Now it can be anything as long as it has is-enabled?
Just making sure this is intentional and not a test regression?

File: source/Classes/1.DscResources/07.nxService.ps1 (hunk 2 above)
Comment (gurrajatwal) [reply source/Classes/1.DscResources/07.nxService.ps1:40]: Before I make this change,

- This resource is written so that if the service doesn't exist, it's compliant.
- The test function is written to return true if there are 0 reasons.

I will have to change the implementation of test if I return a reason. Is that okay?

File: tests/Unit/Classes/DscResources/nxService.tests.ps1 (hunk 3 above)
Comment (gurrajatwal) [reply tests/Unit/Classes/DscResources/nxService.tests.ps1]: This was intentional. I needed to update this mock to cover the new unit test where the service is an empty string. I fixed this in the latest commit so instead of having a vague mock, I have two explicit mocks. I think it's clearer.

File: source/Classes/1.DscResources/07.nxService.ps1 (hunk 2 above)
Comment (kwirkykat) [reply source/Classes/1.DscResources/07.nxService.ps1:40]: So if I say that I want a service Enabled and in the Running state on the machine, but that service doesn't exist on the machine, then we are compliant?
I think as a customer that is not what I would want?
If the customer said that they want the service not Enabled or in like a Disabled state and it doesn't exist, then that seems compliant?
It does seem like this would be a breaking change, but as it is now I think it's broken to say that if the service does not exist, then the resource is always compliant?

Comment (gurrajatwal) [reply source/Classes/1.DscResources/07.nxService.ps1]: Thanks Katie. I updated this to return a reason in some cases. More context in the email I sent. Let me know if we need to discuss.

File: source/Classes/1.DscResources/07.nxService.ps1 [hunk 4]
```diff
@@ -35,18 +35,25 @@ class nxService
...
         $currentState = [nxService]::new()
+        if ([string]::IsNullOrEmpty($nxService.Name))
+        {
+            if ($null -ne $this.Enabled -and -not $this.Enabled -and $null -ne $this.State -and $this.State -eq [nxServiceState]::Stopped)
+            {
+                return $currentState
+            }
+
+            $currentState.Reasons = [Reason]@{
+                Code = '{0}:{0}:NotRunning' -f 'nxService'
```
Comment (kwirkykat) [inline source/Classes/1.DscResources/07.nxService.ps1:46]: This reason should say that the service does not exist. Not just that it isn't running (makes it sound like the service is there and installed, but in a Stopped state)

File: source/Classes/1.DscResources/07.nxService.ps1 [hunk 5]
```diff
@@ -68,7 +75,7 @@ class nxService
         {
             'Enabled'
             {
-                if ($null -ne $this.enabled -and $currentState -and $this.Enabled -ne $currentState.Enabled)
+                if ($null -ne $this.enabled -and $this.Enabled -ne $currentState.Enabled)
```
Comment (kwirkykat) [inline source/Classes/1.DscResources/07.nxService.ps1:78]: nit: enabled should be uppercase

File: source/Classes/1.DscResources/07.nxService.ps1 [hunk 6]
```diff
@@ -84,7 +91,7 @@ class nxService
 
             'State'
             {
-                if ($null -ne $this.State -and $currentState -and $this.State -ne $currentState.State)
+                if ($null -ne $this.State -and $this.State -ne $currentState.State)
```
Comment (kwirkykat) [inline source/Classes/1.DscResources/07.nxService.ps1:94]: Can the state comparison be case insensitive? (ine instead of ne)

File: source/Classes/1.DscResources/07.nxService.ps1 [hunk 7]
```diff
@@ -35,18 +35,25 @@ class nxService
     {
         $nxService = Get-nxService -Name $this.Name
         $currentState = [nxService]::new()
+        if ([string]::IsNullOrEmpty($nxService.Name))
+        {
+            if ($null -ne $this.Enabled -and -not $this.Enabled -and $null -ne $this.State -and $this.State -eq [nxServiceState]::Stopped)
```
Comment (kwirkykat) [inline source/Classes/1.DscResources/07.nxService.ps1:40]: $this.State is a string while we are comparing it to an nxServiceState. Is PowerShell ok with that? Just asking; not sure

File: source/Classes/1.DscResources/07.nxService.ps1 (hunk 4 above)
Comment (gurrajatwal) [reply source/Classes/1.DscResources/07.nxService.ps1]: The command that we are using treats both cases the same.

1. When the resource isn't running (Enabled=False and State=Stopped)
2. When the service doesn't exist and is not installed.

In both cases, "systemctl list-units --type=service --no-legend --all --no-pager \<SERVICENAME\>" will return empty output (at least on Ubuntu 20). I updated the text to "Service ''{0}'' is not running **or does not exist**." Let me know what you think.

File: source/Classes/1.DscResources/07.nxService.ps1 (hunk 7 above)
Comment (gurrajatwal) [reply source/Classes/1.DscResources/07.nxService.ps1:40]: I couldn't find a lot of information online. I found [one stackoverflow thread](https://stackoverflow.com/a/57834292) that says this is how you compare strings with enums. I went back and manually tested again and it's working.

**Review by kwirkykat — APPROVED** [review]

**Comment by eehret** [comment]
@gurrajatwal , @kwirkykat 
Thank you both for working on this.  We are waiting on the release of this fix as well as a number of others, so I'm eager to learn when the next package will be pushed to Powershell Gallery.  Do you have a rough idea when the next version might appear there?
Thanks again!  :)
